File: dotnet/Core/GameX/App/ImportManager.cs
using System;
using System.IO;
using System.Threading.Tasks;

namespace GameX.App;

public static class ImportManager {
    const int MaxDegreeOfParallelism = 8; //1;

    public static async Task ImportAsync(Family family, Resource resource, string filePath, Func<string, bool> match, int from, object option) {
        //foreach (var path in resource.Paths)
        //{
        //    using var arc = family.OpenArchive(resource.Game, new[] { path }) as BinaryArchive;
        //    if (arc == null) throw new InvalidOperationException("Arc not a BinaryArchive");

        //    // import arc
        //    var w = await ImportPakAsync(filePath, from, path, option, arc);
        //}
    }

    static async Task<BinaryWriter> ImportPakAsync(string filePath, int from, string path, object option, BinaryAsset arc) {
        // import arc
        var w = new BinaryWriter(new FileStream(path, FileMode.Create, FileAccess.Write));
        await arc.ImportAsync(w, filePath, from, option, (file, index) => {
            //if ((index % 50) == 0)
            //    Console.WriteLine($"{file.Path}");
        }, (file, message) => {
            Console.WriteLine($"{message}: {file?.Path}");
        });
        return w;
    }

    static async Task ImportAsync(this BinaryAsset source, BinaryWriter w, string filePath, int from = 0, object option = default, Action<FileSource, int> advance = null, Action<FileSource, string> exception = null) {
        // read arc
        if (string.IsNullOrEmpty(filePath) || !Directory.Exists(filePath)) { exception?.Invoke(null, $"Directory Missing: {filePath}"); return; }
        var setPath = Path.Combine(filePath, ".set");
        using (var r = new BinaryReader(File.Open(setPath, FileMode.Open, FileAccess.Read, FileShare.Read))) await ArcBinary.Stream.Read(source, r, "Set");
        var metaPath = Path.Combine(filePath, ".meta");
        using (var r = new BinaryReader(File.Open(setPath, FileMode.Open, FileAccess.Read, FileShare.Read))) await ArcBinary.Stream.Read(source, r, "Meta");
        var rawPath = Path.Combine(filePath, ".raw");
        if (File.Exists(rawPath)) using (var r = new BinaryReader(File.Open(rawPath, FileMode.Open, FileAccess.Read, FileShare.Read))) await ArcBinary.Stream.Read(source, r, "Raw");

        // write header
        if (from == 0) await source.ArcBinary.Write(source, w, "Header");

        // write files
        Parallel.For(0, source.Files.Count, new ParallelOptions { MaxDegreeOfParallelism = MaxDegreeOfParallelism }, async index => {
            var file = source.Files[index].Fix();
            var newPath = Path.Combine(filePath, file.Path);

            // check directory
            var directory = Path.GetDirectoryName(newPath);
            if (string.IsNullOrEmpty(directory) || !Directory.Exists(directory)) { exception?.Invoke(file, $"Directory Missing: {directory}"); return; }

            // insert file
            try {
                await source.ArcBinary.Write(source, w);
                using (var s = File.Open(newPath, FileMode.Open, FileAccess.Read, FileShare.Read)) await source.WriteData(file, s, option);
                advance?.Invoke(file, index);
            }
            catch (Exception e) { ArcBinary.HandleException(file, option, $"Exception: {e.Message}"); }
        });
    }
}

File: rust/core/gamex/src/app/import_manager.rs
// PORT-SOURCE: Core/GameX/App/ImportManager.cs
// PORT-SHA: de2a9d0984627dda
// PORT-STATUS: done
//
// NOT PORTED — there is no working implementation to port.
//
// ============ THE `.meta` SIDECAR IS NEVER READ ==========================
//
//     var setPath = Path.Combine(filePath, ".set");
//     using (var r = new BinaryReader(File.Open(setPath, ...)))
//         await ArcBinary.Stream.Read(source, r, "Set");
//     var metaPath = Path.Combine(filePath, ".meta");
//     using (var r = new BinaryReader(File.Open(setPath, ...)))
//         //                                    ^^^^^^^ setPath again
//         await ArcBinary.Stream.Read(source, r, "Meta");
//
// The second `File.Open` reads **`setPath`**, not `metaPath`. `metaPath` is
// computed and never used, so `.set` is parsed twice — once as a file list and
// once as metadata — and `.meta` is never opened at all.
//
// The consequence is not a crash: `parse_meta` over a `.set` file finds none of
// its section headers and returns an empty result, so every imported archive
// silently loses its compression and encryption flags. `formats/stream.rs`
// documents that parser. **Fix in the C# tree** — one identifier.
//
// The rest of the file is unfinished:
//
//   * **The public entry point `ImportAsync` has an empty body.** Its whole
//     content is a commented-out `foreach (var path in resource.Paths)`, so
//     import does nothing and reports success.
//   * **`MaxDegreeOfParallelism = 8; //1;`** is declared and never read — the
//     trailing `//1;` looks like someone toggling it while debugging.
//   * `ImportPakAsync` creates a `BinaryWriter` over a `FileStream` and returns
//     it without disposing either; the caller receives an open handle with no
//     indication it owns it.
//
// Nothing to translate until the C# does something. Kept as a file so the 1:1
// mapping holds and `sync-check.sh` notices when it grows an implementation.
